packages/vite/src/resolver.ts

```diff
@@ -108,3 +119,75 @@ async function observeDepScan(context: PluginContext, source: string, importer:
   writeStatus(source, result ? 'found' : 'not_found');
   return result;
 }
+
+function idFromResult(result: ResolveIdResult): string | undefined {
+  if (!result) {
+    return undefined;
+  }
+  if (typeof result === 'string') {
+    return cleanUrl(result);
+  }
+  return cleanUrl(result.id);
+}
+
+function hashed(path: string): string {
+  let h = createHash('sha1');
+  return h.update(path).digest('hex').slice(0, 8);
+}
+
+async function maybeCaptureNewOptimizedDep(
+  context: PluginContext,
+  resolver: Resolver,
+  result: ResolveIdResult,
+  notViteDeps: Set<string>
+): Promise<ResolveIdResult> {
+  let foundFile = idFromResult(result);
+  if (!foundFile) {
+    return result;
+  }
+  if (foundFile.startsWith(join(resolver.packageCache.appRoot, 'node_modules', '.vite'))) {
+    debug('maybeCaptureNewOptimizedDep: %s already in vite deps', foundFile);
+    return result;
+  }
+  let pkg = resolver.packageCache.ownerOfFile(foundFile);
+  if (!pkg?.isV2Addon()) {
+    debug('maybeCaptureNewOptimizedDep: %s not in v2 addon', foundFile);
+    return result;
+  }
+  let target = externalName(pkg.packageJSON, explicitRelative(pkg.root, foundFile));
+  if (!target) {
+    debug('maybeCaptureNewOptimizedDep: %s is not exported', foundFile);
+    return result;
+  }
+
+  if (notViteDeps.has(foundFile)) {
+    debug('maybeCaptureNewOptimizedDep: already attmpted %s', foundFile);
+    return result;
+  }
+
+  debug('maybeCaptureNewOptimizedDep: doing re-resolve for %s ', foundFile);
+
+  let jumpRoot = join(tmpdir, 'embroider-vite-jump', hashed(pkg.root));
```

Comment:
Ah, nice trick 
There was an advantage in the previous implementation that it was doing it only for rewritten packages,  so auto-upgraded.

It might be that now this will also optimise other deps.

If a dep is excluded in vite, the default behaviour is to also exclude the dependenciea of it.
If this really does optimise deps of excluded deps i think it would be nice and definitely a candidate for another vite plugin :)

Reply:
> There was an advantage in the previous implementation that it was doing it only for rewritten packages, so auto-upgraded.

Yeah, I think that probably left a bug for native v2 addons, since they also need this behavior but won't get rewritten. Consider your app depends on `addon-one` which depends on `addon-two` and both are native v2 addons so neither is rewritten.

During dev if you add a new usage of a component from `addon-two`, module resolving will look like:

1. defaultResolve runs for `your-app/components/the-new-one` and doesn't find it.
2. fallbackResolve finds it in the mergeMap and retries with `request.alias('addon-two/_app_/components/the-new-one.js').rehome('/path/to/your-app/node_modules/addon-one/package.json')`. The rehome here is chosen to be in a package that can definitely resolve `addon-two`. 
3. defaultResolve finds the file, and the specifier is a bare import, but the fromFile includes `node_modules`, so you get an un-optimized dependency.

> If a dep is excluded in vite, the default behaviour is to also exclude the dependencies of it. Because it's in node_modules.

That's a good point, and another reason why vite's isInNodeModules check is silly. It should really be crawling the package graph to decide where the boundary between not-optimized and optimized packages is.

Yeah, a plugin could fix that problem for everyone, although IMO it would be better to fix it inside vite. I think their package info caches are equivalently powerful to the packageCache in embroider, so could solve the problem without a lot of extra infrastructure.